### DISCUSSION TITLE
emplace throws exception if key already exists

### DISCUSSION THREAD

**Discussion by hoytech** [taocpp/json#58] [Q&A] [answered]
If I have a `tao::json::value` and I call `emplace` on it like this:

    auto pair = output.emplace(key, tao::json::empty_object);

If the key already exists I see this error: `Fatal error: duplicate JSON object key "..."`

Is this expected behaviour? With `std::map` I believe it will succeed and return a pair with an iterator to the existing element and false to indicate no insertion occurred.

Thanks!

**Comment by d-frey** [comment, marked as answer]
Yes, this is the intended behavior.

Our rationale is: A JSON value will always be a hybrid between different things as it is a union type. It has a (partial) map-like interface, a (partial) vector-like interface and a bunch of other methods.

If you use the direct interface of `value`, you'll get some "support" for what we consider reasonable behavior for a JSON value. In this case, you will get an exception if you use `emplace` with a duplicate key. We thought about this in the past and decided that it is the most commonly useful behavior for that method.

If you want the original `map` interface, you can always access the embedded map (given that the value currently stores a JSON object) and use the `std::map`'s methods directly:

    auto result = v.get_object().emplace(...);

(Note that `get_object()` will throw an exception if `v` does not contain an object)

As we haven't published a 1.0.0 version, please let us know if you have different ideas/reasoning about the interface, we are still open to discussions and improvements.

**Comment by hoytech** [comment]
Thanks for the explanation, it sounds reasonable. The `get_object()` approach worked fine for me.

I don't really have any strong preference either way. My only comment is that this wasn't clear to me from the documentation (or maybe I wasn't looking in the right spots?). Since this behaviour is different from how `emplace` usually works, I think it would be helpful to have this described more prominently in the docs.

I appreciate the prompt response, and the library in general!

**Comment by d-frey** [comment]
Thank you, and you are right that the documentation needs some improvement. We hope to get there once we will approach 1.0.0.
